### PR TITLE
Move some code into models/git

### PR DESCRIPTION
**Pull request by lunny**
depends on #19711 to move lfs related code.

### CODE REVIEW

File: models/git/branches.go [hunk 1]
```diff
@@ -159,71 +160,9 @@ func isUserOfficialReviewer(ctx context.Context, protectBranch *ProtectedBranch,
...
-		return true
-	}
-
-	return has
-}
-
 // MergeBlockedByOutdatedBranch returns true if merge is blocked by an outdated head branch
-func (protectBranch *ProtectedBranch) MergeBlockedByOutdatedBranch(pr *PullRequest) bool {
-	return protectBranch.BlockOnOutdatedBranch && pr.CommitsBehind > 0
+func (protectBranch *ProtectedBranch) MergeBlockedByOutdatedBranch(commitsBehind int) bool {
```
Comment (KN4CK3R) [inline models/git/branches.go]: The remove of the `PullRequest` parameter is to prevent an import cycle (`MergeBlockedByProtectedFiles` too)? The old version looked more "useful".

Or move them into `models/branch.go` too.

File: models/git/branches_test.go [hunk 2]
```diff
@@ -2,12 +2,14 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
-package models
+package git_test
```
Comment (KN4CK3R) [inline models/git/branches_test.go:5]: Why not just `git`?

File: models/git/lfs.go [hunk 3]
```diff
@@ -19,6 +20,83 @@ import (
 	"xorm.io/builder"
 )
 
+//.____   ____________________
+//|    |  \_   _____/   _____/
+//|    |   |    __) \_____  \
+//|    |___|     \  /        \
+//|_______ \___  / /_______  /
+//        \/   \/          \/
+
```
Comment (KN4CK3R) [inline models/git/lfs.go]:
```suggestion
```

File: routers/web/repo/repo.go [hunk 4]
```diff
@@ -599,6 +599,13 @@ func SearchRepo(ctx *context.Context) {
 
 	results := make([]*api.Repository, len(repos))
 	for i, repo := range repos {
+		if err = repo.GetOwner(ctx); err != nil {
+			ctx.JSON(http.StatusInternalServerError, api.SearchError{
+				OK:    false,
+				Error: err.Error(),
+			})
+			return
+		}
```
Comment (KN4CK3R) [inline routers/web/repo/repo.go]: Why is this needed? `repo_model.SearchRepository(opts)` initializes the owner property.

File: models/git/branches_test.go (hunk 2 above)
Comment (lunny) [reply models/git/branches_test.go:5]: Because the test may depends on some sub packages which depends on `git`. Since all testing functions are public. A standalone test package is more flexible.

File: models/git/branches.go [hunk 5]
```diff
@@ -252,13 +191,13 @@ func getFilePatterns(filePatterns string) []glob.Glob {
...
 // MergeBlockedByProtectedFiles returns true if merge is blocked by protected files change
-func (protectBranch *ProtectedBranch) MergeBlockedByProtectedFiles(pr *PullRequest) bool {
+func (protectBranch *ProtectedBranch) MergeBlockedByProtectedFiles(changedProtecedFiles []string) bool {
 	glob := protectBranch.GetProtectedFilePatterns()
 	if len(glob) == 0 {
 		return false
 	}
 
-	return len(pr.ChangedProtectedFiles) > 0
+	return len(changedProtecedFiles) > 0
```
Comment (delvh) [inline models/git/branches.go]:
```suggestion
	return len(changedProtectedFiles) > 0
```

File: models/git/branches.go [hunk 6]
```diff
@@ -252,13 +191,13 @@ func getFilePatterns(filePatterns string) []glob.Glob {
 }
 
 // MergeBlockedByProtectedFiles returns true if merge is blocked by protected files change
-func (protectBranch *ProtectedBranch) MergeBlockedByProtectedFiles(pr *PullRequest) bool {
+func (protectBranch *ProtectedBranch) MergeBlockedByProtectedFiles(changedProtecedFiles []string) bool {
```
Comment (delvh) [inline models/git/branches.go]:
```suggestion
func (protectBranch *ProtectedBranch) MergeBlockedByProtectedFiles(changedProtectedFiles []string) bool {
```

File: models/branch.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	}
+	return GetGrantedApprovalsCount(ctx, protectBranch, pr) >= protectBranch.RequiredApprovals
+}
+
+// GetGrantedApprovalsCount returns the number of granted approvals for pr. A granted approval must be authored by a user in an approval whitelist.
+func GetGrantedApprovalsCount(ctx context.Context, protectBranch *git_model.ProtectedBranch, pr *PullRequest) int64 {
+	sess := db.GetEngine(ctx).Where("issue_id = ?", pr.IssueID).
+		And("type = ?", ReviewTypeApprove).
+		And("official = ?", true).
+		And("dismissed = ?", false)
```
Comment (delvh) [inline models/branch.go:28]: I've seen this code block twice (potentially even thrice) now: Here, in 46-49 below, and perhaps even in 65-67.
What about extracting a private function that takes as parameters `ctx context.Context, pr *PullRequest, reviewType ReviewType` and returns the engine after executing all these statements?

Comment (lunny) [reply models/branch.go:28]: I would leave them there for other PRs since I just move functions in this PRs.

**Comment by lunny** [comment]
@KN4CK3R @delvh Both done.

**Review by delvh — APPROVED** [review]

**Review by KN4CK3R — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-m work
